### ISSUE TITLE
Multi/single rail in the PSM2 provider of the OFI MTL provider

### ISSUE THREAD

**Issue by jsquyres** [open-mpi/ompi#6256]
Similar to #6255, a user asked me if there's a way to force single and/or multi-rail for the PSM2 provider in the OFI MTL.

Specifically: I think the default is multi-rail.  Is there a way to specify which PSM2 device(s) that the OFI MTL should use?

This was asked in the context of v4.0.0, but I assume the answer is the same on all the release branches.

**Comment by aravindksg** [comment]
PSM2 library handles the support for multi-rail which is off by default (for single rank jobs). For multi-rank jobs, the default PSM2 behavior depends on the affinity settings of the MPI process as PSM2 defaults to using the HFI local to the MPI process. The multi-rail settings can be modified using the following environment variables-

- PSM2_MULTIRAIL=[0,1,2] (0=Disabled, 1=Enable across all HFIs in the system, 2=Enable multirail within a specific NUMA node)
- PSM2_MULTIRAIL_MAP=unit:port,unit:port...

To restrict PSM2 to use a single HFI, use the following environment variable-
- HFI_UNIT=[value]

More details can be found on the PSM2 Programmer's Guide and Omni-Path Fabric Performance Tuning Guide. Full documentation here:
https://www.intel.com/content/dam/support/us/en/documents/network-and-i-o/fabric-products/Intel_Omni-Path_Software_publications_10_8.zip

**Comment by jsquyres** [comment]
Great -- thanks!

Just to clarify: you said that multi-rail is off by default for "single-rank jobs" -- do you mean single *process* jobs?  I would imagine single/multi-rail is not very important for an MPI job that is a single process.

Regardless, can you make an ompi-www FAQ item about this?  I think it would be most helpful to see this information in an MPI context -- e.g., `mpirun -x PSM2_MULTIRAIL=...`.

**Comment by aravindksg** [comment]
Ah, sorry about that, I ended up quoting the PSM2 programmer's guide there (which as you point out might require edits for clarity). So, essentially, I meant the number of local ranks in a node ("MPI_LOCALNRANKS"). If this value is 1, then multi-rail is off by deafult. Otherwise, it depends on the user's affinity settings.

Sure, I can work on an FAQ entry for this. I think this falls under the "Running MPI jobs" page?

**Comment by jsquyres** [comment]
It might be better under the "Tuning" section -- we generally have a page for each network type.

**Comment by aravindksg** [comment]
Correction to my earlier answer as I mixed between Multi-Rail and Multi-HFI scenarios.

- Multi-Rail feature in PSM2 allows a process to use multiple HFIs to transfer a message to improve message bandwidth. This is off by default and can be controlled by `PSM2_MULTIRAIL` and `PSM2_MULTIRAIL_MAP` variables mentioned above.
- Multi-HFI support is intended to describe the use of multiple HFIs in a system among MPI ranks local to a node in order to load-balance the hardware resources.

I have added an FAQ entry and clarified above scenarios here: https://github.com/open-mpi/ompi-www/pull/233
